packages/dd-trace/src/encode/0.4.js

```diff
@@ -319,4 +339,72 @@ class AgentEncoder {
   }
 }
 
+function formatSpanEvents (span) {
+  for (const spanEvent of span.span_events) {
+    if (spanEvent.attributes) {
+      for (const [key, value] of Object.entries(spanEvent.attributes)) {
+        const newValue = convertSpanEventAttributeValues(key, value)
+        if (newValue !== undefined) spanEvent.attributes[key] = newValue
+        else delete spanEvent.attributes[key] // delete from attributes if undefined
+      }
+      if (Object.entries(spanEvent.attributes).length === 0) {
+        delete spanEvent.attributes
+      }
+    }
+  }
+}
+
+function convertSpanEventAttributeValues (key, value, depth = 0) {
+  if (typeof value === 'string') {
+    return {
+      type: 0,
+      string_value: value
+    }
+  } else if (typeof value === 'boolean') {
+    return {
+      type: 1,
+      bool_value: value
+    }
+  } else if (Number.isInteger(value)) {
+    return {
+      type: 2,
+      int_value: value
+    }
+  } else if (typeof value === 'number') {
+    return {
+      type: 3,
+      double_value: value
+    }
+  } else if (Array.isArray(value)) {
+    if (depth === 0) {
+      const convertedArray = value
+        .map((val) => convertSpanEventAttributeValues(key, val, 1))
+        .filter((convertedVal) => convertedVal !== undefined)
+
+      // Only include array_value if there are valid elements
+      if (convertedArray.length > 0) {
+        return {
+          type: 4,
+          array_value: convertedArray
+        }
+      } else {
+      // If all elements were unsupported, return undefined
+        return undefined
+      }
+    } else {
+      log.warn(
```

Comment:
What if this happens on every span? It will be way too verbose for the WARN level. Could it be memoized somehow? I think there is memoization in some other places in dd-trace which hopefully can be reused.

Reply:
Lowered to DEBUG log level and added memoization, including tests to ensure memoization works.